File: app/src/com/ternaryop/photoshelf/db/Birthday.java
package com.ternaryop.photoshelf.db;

import java.io.Serializable;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

import android.provider.BaseColumns;

public class Birthday implements BaseColumns, Serializable {
	private static final long serialVersionUID = 5887665998065237345L;
    private long id;
	private String name;
	private Date birthDate;
	private String tumblrName;

	public static final SimpleDateFormat ISO_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd", Locale.US);
	
	public Birthday() {
	}

	public Birthday(String name, Date birthDate, String tumblrName) {
		this.tumblrName = tumblrName;
		this.name = name.toLowerCase(Locale.US);
		this.birthDate = birthDate;
	}

	public Birthday(String name, String birthDate, String tumblrName) throws ParseException {
		this(name, birthDate == null || birthDate.trim().length() == 0 ? null : ISO_DATE_FORMAT.parse(birthDate), tumblrName);
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name.toLowerCase(Locale.US);
	}

	public Date getBirthDate() {
		return birthDate;
	}

    public Calendar getBirthDateCalendar() {
        Calendar c = Calendar.getInstance();
        c.setTime(birthDate);
        return c;
    }

	public void setBirthDate(Date birthDate) {
		this.birthDate = birthDate;
	}

	public String getTumblrName() {
		return tumblrName;
	}

	public void setTumblrName(String tumblrName) {
		this.tumblrName = tumblrName;
	}
	
	@Override
	public String toString() {
		return (birthDate == null ? "" : ISO_DATE_FORMAT.format(birthDate) + " ") + name;
	}

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/db/Birthday.kt
package com.ternaryop.photoshelf.db

import android.provider.BaseColumns
import com.ternaryop.photoshelf.util.date.year
import java.io.Serializable
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

class Birthday : BaseColumns, Serializable {
    var id: Long = 0
    var name: String
    set(value) {
        field = value.toLowerCase(Locale.US)
    }
    var birthDate: Calendar? = null
    var tumblrName: String

    val age: Int
        get() = Calendar.getInstance().year - birthYear

    val birthYear: Int
        get() = birthDate?.year ?: 0

    constructor(name: String, birthDate: Calendar?, tumblrName: String) {
        this.tumblrName = tumblrName
        this.name = name.toLowerCase(Locale.US)
        this.birthDate = birthDate
    }

    @Throws(ParseException::class)
    constructor(name: String, birthDate: String?, tumblrName: String) :
            this(name, if (birthDate == null || birthDate.isBlank()) null else fromIsoFormat(birthDate), tumblrName)

    override fun toString(): String {
        return (if (birthDate == null) "" else ISO_DATE_FORMAT.format(birthDate!!.time) + " ") + name
    }

    companion object {
        private const val serialVersionUID = 5887665998065237345L

        private val ISO_DATE_FORMAT = SimpleDateFormat("yyyy-MM-dd", Locale.US)

        fun fromIsoFormat(isoDateString: String): Calendar {
            val c = Calendar.getInstance()
            c.time = ISO_DATE_FORMAT.parse(isoDateString)
            return c
        }

        fun toIsoFormat(date: Calendar): String = ISO_DATE_FORMAT.format(date.time)
    }
}